From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Sun, 25 Dec 2022 13:35:52 +0100
Subject: [PATCH] dt-bindings: nand: meson: add support for more SoCs

Older Amlogic SoCs have a slightly different integration of the NFC
(NAND flash controller) than the new ones (GXL, GXM, AXG).

On GXL, AXG and newer the "NAND device clock" is shared with sd_emmc_c.
This requires muxing the signal of that clock between the sd_emmc_c
controller and the NAND controller. For these newer SoCs there's a
second "reg" entry with the sd_emmc_c clock registers.

Older SoCs (Meson8, Meson8b, Meson8m2 and GXBB) have a dedicated "NAND
device clock". Thus we don't need the second "reg" entry for muxing the
NAND clock.

The ECC configuration on the older SoCs (Meson, Meson8b, Meson8m2 and
GXBB) is identical to GXL (which means: ECC strengths 8, 16, 24, 30,
40, 50 and 60 are supported).

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
---
 .../bindings/mtd/amlogic,meson-nand.yaml      | 48 ++++++++++++++++---
 1 file changed, 41 insertions(+), 7 deletions(-)

diff --git a/Documentation/devicetree/bindings/mtd/amlogic,meson-nand.yaml b/Documentation/devicetree/bindings/mtd/amlogic,meson-nand.yaml
index 28fb9a7dd7..a0b478adf3 100644
--- a/Documentation/devicetree/bindings/mtd/amlogic,meson-nand.yaml
+++ b/Documentation/devicetree/bindings/mtd/amlogic,meson-nand.yaml
@@ -6,19 +6,21 @@ $schema: http://devicetree.org/meta-schemas/core.yaml#
 
 title: Amlogic NAND Flash Controller (NFC) for GXBB/GXL/AXG family SoCs
 
-allOf:
-  - $ref: nand-controller.yaml
-
 maintainers:
   - liang.yang@amlogic.com
 
 properties:
   compatible:
     enum:
+      - amlogic,meson8-nfc
+      - amlogic,meson8b-nfc
+      - amlogic,meson8m2-nfc
+      - amlogic,meson-gxbb-nfc
       - amlogic,meson-gxl-nfc
       - amlogic,meson-axg-nfc
 
   reg:
+    minItems: 1
     maxItems: 2
 
   reg-names:
@@ -33,9 +35,13 @@ properties:
     minItems: 2
 
   clock-names:
-    items:
-      - const: core
-      - const: device
+    oneOf:
+      - items:
+          - const: core
+          - const: device
+      - items:
+          - const: core
+          - const: nand
 
 patternProperties:
   "^nand@[0-7]$":
@@ -55,9 +61,18 @@ patternProperties:
         enum: [8, 16, 24, 30, 40, 50, 60]
         description: |
           The ECC configurations that can be supported are as follows.
-            meson-gxl-nfc 8, 16, 24, 30, 40, 50, 60
+            meson8/8b/8m2/gxbb/gxl-nfc 8, 16, 24, 30, 40, 50, 60
             meson-axg-nfc 8
 
+allOf:
+  - $ref: nand-controller.yaml
+
+  - if:
+      properties:
+        compatible:
+          contains:
+            const: amlogic,meson-axg-nfc
+
 required:
   - compatible
   - reg
@@ -85,6 +100,25 @@ examples:
       #address-cells = <1>;
       #size-cells = <0>;
 
+      nand@0 {
+        reg = <0>;
+      };
+    };
+  - |
+    #include <dt-bindings/interrupt-controller/arm-gic.h>
+    nand-controller@48600 {
+      compatible = "amlogic,meson8b-nfc";
+      reg = <0x48600 0x24>;
+      interrupts = <GIC_SPI 34 IRQ_TYPE_EDGE_RISING>;
+      clocks = <&nand_pclk>, <&nand_clock>;
+      clock-names = "core", "nand";
+
+      pinctrl-0 = <&nand_pins>;
+      pinctrl-names = "default";
+
+      #address-cells = <1>;
+      #size-cells = <0>;
+
       nand@0 {
         reg = <0>;
       };
